### PR TITLE
Add corev1.ConfigMap client functions and types

### PR DESCRIPTION
**Pull request by axbarsan**
Towards https://github.com/giantswarm/giantswarm/issues/16717

This PR adds client-side types and functions for dealing with MAPI `ConfigMap` resources.

### CODE REVIEW

File: src/model/services/mapi/corev1/types.ts [hunk 1]
```diff
@@ -23,3 +23,14 @@ export interface IServiceAccount {
...
   automountServiceAccountToken?: boolean;
 }
+
+export const ConfigMap = 'ConfigMap';
+
+export interface IConfigMap {
+  apiVersion: 'v1';
+  kind: typeof ConfigMap;
+  metadata: metav1.IObjectMeta;
+  immutable?: boolean;
```
Comment (kopiczko) [inline src/model/services/mapi/corev1/types.ts:33]: TIL CM has immutable field

File: src/model/services/mapi/corev1/createConfigMap.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    isCore: true,
+    kind: 'configmaps',
+    namespace: configMap.metadata.namespace!,
+  });
+
+  return createResource<IConfigMap>(
+    client,
+    auth,
+    url.toString(),
+    (configMap as unknown) as Record<string, unknown>
```
Comment (kopiczko) [inline src/model/services/mapi/corev1/createConfigMap.ts:24]: This is a bit unclear to me.

**Review by kopiczko — APPROVED** [review]
Looks sane

File: src/model/services/mapi/corev1/createConfigMap.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import { IOAuth2Provider } from 'lib/OAuth2/OAuth2';
+import { IHttpClient } from 'model/clients/HttpClient';
+import * as k8sUrl from 'model/services/mapi/k8sUrl';
```
Comment (kopiczko) [inline src/model/services/mapi/corev1/createConfigMap.ts:3]: I guess there are reasons to not use some external k8s libs?

File: src/model/services/mapi/corev1/types.ts (hunk 1 above)
Comment (kubasobon) [reply src/model/services/mapi/corev1/types.ts:33]: **+1**
> Starting from v1.19, you can add an immutable field to a ConfigMap definition to create an immutable ConfigMap.

https://kubernetes.io/docs/concepts/configuration/configmap/#configmap-immutable

File: src/model/services/mapi/corev1/createConfigMap.ts (hunk 2 above)
Comment (axbarsan) [reply src/model/services/mapi/corev1/createConfigMap.ts]: The type I used was for trying to make sure that we use a data structure derived from a map here, and avoid using `any`, at the same time. But seems like `any` was actually the way to go

File: src/model/services/mapi/corev1/createConfigMap.ts (hunk 3 above)
Comment (axbarsan) [reply src/model/services/mapi/corev1/createConfigMap.ts:3]: The library scene for client-side k8s is pretty much unexistent at this time. Pretty much everything there is out there is for running on the server side.

We had some automation setup for generating all the client-side types and methods from our `apiextensions` repo [in here](https://github.com/giantswarm/giantswarm-cp-client). However, it was very brittle (something would always crash when adding new types), and the generated code was a bit clunky. Due to how it was built, we couldn't strip unused code at build time, so we had a huge bundle.

We reverted to using the K8s API as a regular REST API for now, and created a few utilities, such as `k8sUrl` to make it a bit easier

Comment (kopiczko) [reply src/model/services/mapi/corev1/createConfigMap.ts:3]: How about this https://github.com/kubernetes-client/javascript ?

Comment (axbarsan) [reply src/model/services/mapi/corev1/createConfigMap.ts:3]: That's only for the server-side.
https://github.com/kubernetes-client/javascript/issues/165

I tried porting it to the browser at some point, but there were so many Node.js native libraries to create, that I gave up eventually

Comment (kopiczko) [reply src/model/services/mapi/corev1/createConfigMap.ts:3]: Alright I see what you mean by "client-side" now :)

**Pull request merged**